packages/cli-vector/src/generalization/generalization.ts

```diff
@@ -32,87 +30,118 @@ export async function generalize(input: URL, output: URL, options: VectorCreatio
     if (line === '') continue;
     inputCount++;
-    // For simplify Duplicate feature for each zoom level with different tolerance
+    // For simplify, duplicate feature for each zoom level with different tolerance
     if (simplify != null) {
       for (const s of simplify) {
-        const feature = tag(options, line, s);
+        const feature = tag(options, line, s, logger);
         if (feature == null) continue;
+
+        writeStream.write(JSON.stringify(feature) + '\n');
         outputCount++;
-        features.push(JSON.parse(feature) as VectorGeoFeature);
       }
     } else {
-      const feature = tag(options, line);
+      const feature = tag(options, line, null, logger);
       if (feature == null) continue;
-      outputCount++;
-      features.push(JSON.parse(feature) as VectorGeoFeature);
-    }
-  }
-  options.layer.metrics = { input: inputCount, output: outputCount };
 
-  if (features.length > 0) {
-    const writeStream = createWriteStream(output);
-    for (const feature of features) {
       writeStream.write(JSON.stringify(feature) + '\n');
+      outputCount++;
     }
-    writeStream.close();
-    return output;
   }
-  return undefined;
+
+  await new Promise((resolve) => {
+    writeStream.close(resolve);
+  });
+
+  // special handling for the gazatteer (place_labels) dataset
+  // issues: bypasses the 'simplify' and 'removeAttributes' operations of the 'tag' function
+  // if (options.layer.id === '51154') {
+  //   logger.info({}, 'Special handling for the gazatteer (place_labels) dataset');
+  //   features = Array.from(PlaceLabelsFeatures.values());
+  //   outputCount = features.length;
+  // }
+
+  const metrics: Metrics = {
+    input: inputCount,
+    output: outputCount,
+  };
+
+  logger.info({ inputCount, outputCount }, 'Generalize:End');
+  return metrics;
 }
 
 /**
  * Tag feature for layer
  */
-function tag(options: VectorCreationOptions, line: string, simplify?: Simplify): string | undefined {
-  const feature = JSON.parse(line) as VectorGeoFeature;
-
-  feature['tippecanoe'] = {
-    layer: options.name,
-    minzoom: options.layer.style.minZoom,
-    maxzoom: options.layer.style.maxZoom,
-  };
+function tag(
+  options: VectorCreationOptions,
+  line: string,
+  simplify: Simplify | null,
+  logger: LogType,
+): VectorGeoFeature | null {
+  logger.debug({}, 'Tag:Start');
+  const feature = VectorGeoFeatureSchema.parse({
+    ...JSON.parse(line),
+    tippecanoe: {
+      layer: options.name,
+      minzoom: options.layer.style.minZoom,
+      maxzoom: options.layer.style.maxZoom,
+    },
+  });
 
+  // copy the stac json's tags to the feature (i.e. 'kind')
   Object.entries(options.layer.tags).forEach(([key, value]) => (feature.properties[key] = value));
 
-  // TODO: Add support of special tagging logics.
+  // adjust the feature's metadata and properties
+  const modifiedFeature = modifyFeature(feature, options, logger);
+  if (modifiedFeature == null) {
+    logger.debug({}, 'Tag:End');
```

Comment:
I feel like these logs are not needed with not much information in it. And might dumping terminal a lot if we goes with debug mode.